File: sms-api-java/apigateway/src/main/java/com/example/mydemoproject/apigateway/config/AuthenticationFilter.java
package com.example.mydemoproject.apigateway.config;

import jakarta.validation.constraints.NotNull;
import jakarta.ws.rs.client.Client;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.cloud.client.discovery.DiscoveryClient;
import org.springframework.cloud.context.config.annotation.RefreshScope;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.GatewayFilterChain;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

@AllArgsConstructor
@RefreshScope
@Component
@Configuration
@ConditionalOnProperty(value = "gateway.expose-unsecured-urls", havingValue = "true", matchIfMissing = false)
@Slf4j
public class AuthenticationFilter implements GatewayFilter {

    private final @NotNull ApiGatewayProperties gatewayProperties;

    private final Logger logger = LoggerFactory.getLogger(AuthenticationFilter.class);

    @Autowired
    DiscoveryClient discoveryClient;

    @Autowired
    WebClient.Builder webClientBuilder;

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
        ServerHttpRequest request = exchange.getRequest();

        if (isSecured(request)) {
            if (this.isAuthMissing(request))
                return this.onError(exchange, "Authorization header is missing in request.", HttpStatus.UNAUTHORIZED);

            String authHeader = getAuthHeader(request);
            if (!authHeader.isEmpty()) {
                String jwtToken = authHeader.replace("Bearer ", "");
                return validateToken(exchange, chain, jwtToken);
            } else {
                return this.onError(exchange, "Invalid token provided", HttpStatus.UNAUTHORIZED);
            }

        }

        return chain.filter(exchange);
    }

    public boolean isSecured(ServerHttpRequest request) {
        return gatewayProperties.getUnsecuredUrls()
                .stream()
                .noneMatch(uri -> request.getURI().getPath().contains(uri));
    }

    private Mono<Void> onError(ServerWebExchange exchange, String err, HttpStatus httpStatus) {
        ServerHttpResponse response = exchange.getResponse();
        response.setStatusCode(httpStatus);
        return response.setComplete();
    }

    private Mono<Void> validateToken(ServerWebExchange exchange, GatewayFilterChain chain, String token) {
        return webClientBuilder
                .build()
                .post()
                .uri("http://auth-service/api/auth/validate")
                .bodyValue(token)
                .exchangeToMono(
                        clientResponse -> {
                            if(clientResponse.statusCode().isError()) {
                                return onError(exchange, "", HttpStatus.resolve(clientResponse.statusCode().value()));
                            } else {
                                return chain.filter(exchange);
                            }
                        }
                );
    }

    private String getAuthHeader(ServerHttpRequest request) {
        return request.getHeaders().getOrEmpty("Authorization").get(0);
    }

    private boolean isAuthMissing(ServerHttpRequest request) {
        return !request.getHeaders().containsKey("Authorization");
    }
}

File: sms-api-kotlin/apigateway/src/main/kotlin/com/example/apigateway/config/AuthenticationFilter.kt
package com.example.apigateway.config

import jakarta.validation.constraints.NotNull
import lombok.AllArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.cloud.client.discovery.DiscoveryClient
import org.springframework.cloud.context.config.annotation.RefreshScope
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.GatewayFilterChain
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpStatus
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.ClientResponse
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono


@AllArgsConstructor
@RefreshScope
@Component
@Configuration
@ConditionalOnProperty(value = ["gateway.expose-unsecured-urls"], havingValue = "true", matchIfMissing = false)
@Slf4j
class AuthenticationFilter : GatewayFilter {
    private val gatewayProperties: @NotNull ApiGatewayProperties? = null
    private val logger = LoggerFactory.getLogger(AuthenticationFilter::class.java)

    @Autowired
    var discoveryClient: DiscoveryClient? = null

    @Autowired
    var webClientBuilder: WebClient.Builder? = null
    override fun filter(exchange: ServerWebExchange, chain: GatewayFilterChain): Mono<Void>? {
        val request = exchange.request
        if (isSecured(request)) {
            if (isAuthMissing(request)) return onError(
                exchange,
                "Authorization header is missing in request.",
                HttpStatus.UNAUTHORIZED
            )
            val authHeader = getAuthHeader(request)
            return if (!authHeader.isEmpty()) {
                val jwtToken = authHeader.replace("Bearer ", "")
                validateToken(exchange, chain, jwtToken)
            } else {
                onError(exchange, "Invalid token provided", HttpStatus.UNAUTHORIZED)
            }
        }
        return chain.filter(exchange)
    }

    fun isSecured(request: ServerHttpRequest): Boolean {
        return gatewayProperties!!.unsecuredUrls
            .stream()
            .noneMatch { uri -> request.uri.path.contains(uri) }
    }

    private fun onError(exchange: ServerWebExchange, err: String, httpStatus: HttpStatus?): Mono<Void> {
        val response = exchange.response
        response.setStatusCode(httpStatus)
        return response.setComplete()
    }

    private fun validateToken(exchange: ServerWebExchange, chain: GatewayFilterChain, token: String): Mono<Void>? {
        return webClientBuilder
            ?.build()
            ?.post()
            ?.uri("http://auth-service/api/auth/validate")
            ?.bodyValue(token)
            ?.exchangeToMono { clientResponse: ClientResponse ->
                if (clientResponse.statusCode().isError) {
                    return@exchangeToMono onError(
                        exchange,
                        "",
                        HttpStatus.resolve(clientResponse.statusCode().value())
                    )
                } else {
                    return@exchangeToMono chain.filter(exchange)
                }
            }
    }

    private fun getAuthHeader(request: ServerHttpRequest): String {
        return request.headers.getOrEmpty("Authorization")[0]
    }

    private fun isAuthMissing(request: ServerHttpRequest): Boolean {
        return !request.headers.containsKey("Authorization")
    }
}
